chain/deals/request_validation.go

```diff
@@ -0,0 +1,229 @@
+package deals
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"reflect"
+
+	"github.com/ipfs/go-cid"
+	datastore "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"go.uber.org/fx"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/datatransfer"
+	"github.com/filecoin-project/lotus/node/modules/dtypes"
+)
+
+var (
+	// ErrWrongVoucherType means the voucher was not the correct type can validate against
+	ErrWrongVoucherType = errors.New("cannot validate voucher type")
+
+	// ErrNoPushAccepted just means clients do not accept pushes for storage deals
+	ErrNoPushAccepted = errors.New("Client should not receive data for a storage deal")
+
+	// ErrNoPullAccepted just means providers do not accept pulls for storage deals
+	ErrNoPullAccepted = errors.New("Provider should not send data for a storage deal")
+
+	// ErrNoDeal means no active deal was found for this vouchers proposal cid
+	ErrNoDeal = errors.New("No deal found for this proposal")
+
+	// ErrWrongPeer means that the other peer for this data transfer request does not match
+	// the other peer for the deal
+	ErrWrongPeer = errors.New("Data Transfer peer id and Deal peer id do not match")
+
+	// ErrWrongPiece means that the pieceref for this data transfer request does not match
+	// the one specified in the deal
+	ErrWrongPiece = errors.New("Base CID for deal does not match CID for piece")
+
+	// ErrInacceptableDealState means the deal for this transfer is not in a deal state
+	// where transfer can be performed
+	ErrInacceptableDealState = errors.New("Deal is not a in a state where deals are accepted")
```

Comment:
Nit: sentinel error strings defined like those above prevent the message including useful details such as the piece or deal in question. Exporting the values also encourages dependencies and encourages making comparisons on the error value, which add fragile coupling.

It could be considered a style thing (tho I think it has technical consequences) and so matching the existing style is ok if that's the motivation. But see https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully for a more thorough argument.

Reply:
Also using `xerrors.New/Errorf` includes caller frame information, which is really useful when debugging

Reply:
Sentinel errors are really nice for use with errors.Is, you can always wrap to add extra context (as is common throughout the rest of the codebase)